vpr/src/timing/timing_graph_builder.cpp

```diff
@@ -1,3 +1,196 @@
+ * from the provided AtomNetlist.
+ *
+ * The Timing Graph is a directed acyclic graph (DAG) consisting of nodes and edges:
+ *    - Nodes: represent netlist pins (tatum::IPIN, tatum::OPIN) and logical sources/
+ *      sinks ike primary inputs/outpus, flip-flops and clock generators (tatum::SOURCE,
+ *      tatum::SINK).
+ *    - Edges: represent the timing dependences between nodes in the timing graph. 
+ *      These correspond to net connections, combinational paths inside primitives, 
+ *      clock-to-q, and setup/hold constraints for sequential elements.
+ *
+ * The timing graph is constructed by first creating timing sub-graphs corresponding to 
+ * each primitive block in the AtomNetlist. Each sub-graph contains the timing graph 
+ * nodes required to represent the primitive, and all internal timing graph edges (i.e.
+ * those completely contained within the primitive).
+ *
+ * Next timing graph edges corresponding to nets in the netlist (i.e. connections between 
+ * netlist primitives) are created to "stich" the sub-graph together. This results in a
+ * timing graph corresponding to the netlist.
+ *
+ *
+ * Modelling Primitive Combinational and Sequetial Logic as a Timing Graph
+ * -----------------------------------------------------------------------
+ *
+ * Consider the architectural primitive block below, which contains two 
+ * sequential elements 'A' and 'B' (controlled by the primitive input 
+ * pin 'clk'), and two clouds of combinational logic 'C' and 'D'.
+ *
+ * The combinational logic 'D' is driven by primtive input 'e' and 
+ * drives primitive output pin 'g'.
+ *
+ * The combinational logic 'C' is driven by the sequential element 'A' and
+ * primitive input pin 'e'; it then drives the input of the sequential 
+ * element 'B'.
+ *
+ * Sequentail element 'A' is driven by primitive input 'f', and sequential 
+ * element 'B' drives primitive output pin 'h'.
+ *
+ *           +---------------------------------+
+ *           |                                 |
+ *           |                         ---     |
+ *           |                        /   \    |
+ *     e --->-----------------+----->|  D  |-->|--> g
+ *           |                |       \   /    |   
+ *           |                |        ---     |
+ *           |                v                |
+ *           |               ---               |
+ *           |   +---+      /   \      +---+   |
+ *     f --->|-->| A |---->|  C  |---->| B |-->|--> h
+ *           |   |   |      \   /      |   |   |
+ *           |   |   |       ---       |   |   |
+ *           |   +-^-+                 +-^-+   |
+ *           |     |                     |     |
+ *           |     |                     |     |
+ *   clk --->|---------------------------+     |
+ *           |                                 |
+ *           +---------------------------------+
+ *
+ *  As a result there are the following "timing sub-paths" within this primitive:
+ *
+ *  1) 'e' -> 'D' -> 'g'
```

Comment:
I think it would be useful to expand the types of edges here?  If I understand e -> and D -> g, those are both simple propigation delay?  clk -> A is a clk-to-Q edge (e.g. launch delay).

Should `'f' -> 'A'` be a `'f' -> 'A.D', and `'A' -> 'C'` be `'A.Q' -> 'C'`, etc.?  Because A is actually 3 ports, A.D, A.clk, and A.Q.  There is not edge from A.D to A.Q at all.  Only `A.D` captured from `A.clk`, and `A.Q` launched from `A.clk`.  Yes?